ruby/README.md

```diff
@@ -827,6 +824,179 @@ class TestClass
+        Correction.create!
+        write_to_logs("success!")
+      elsif response["code"] == 403
+        write_to_logs("unauthorized")
+      end
+    end
+
+    # Problem: Unnecessary Public Method makes the purpose of the class unclear
+    def write_to_logs(message)
+      # Problem: Constructing a custom logger might require changes
+      # as soon as we need to improve logging across the application.
+      Log4r::Logger.new("Application Log").info(message)
+    end
+
+    private
+    # Problem: Unencapsulated API behavior -- this class might require changes
+    # as soon as this third party API changes.
+    def call_third_party_api
+      response = Http.get('www.invoca.net').response
+      JSON.parse(response)
+    end
+  end
+  ```
+
+  We've extracted logging to a module which can be reused and have established a contract between this class and the module. So long as `log_info` is maintained, Logging can be freely extended. We also marked `log_info` as protected, which ensures that future developers are not able to violate the encapsulation of CorrectionProcessor and our Logging module by using CorrectionProcessor for logging for other classes.
+  ```ruby
+  # Good
+  module Logging
+    protected
+
+    def log_info(message)
+      logger.info(message)
+    end
+
+    private
+
+    def logger
+      @logger ||= Log4r::Logger.new
+    end
+  end
+
+  class CorrectionProcessor
+    # Said aloud: "This class processes a correction"
+    include Logging
+
+    def process
+      if invoca_api_call.success?
+        Correction.create!
+        log_info("success!")
+      elsif invoca_api_call.unauthorized?
+        log_info("unauthorized")
+      end
+    end
+
+    # continued below..
+  ```
+
+  We've extracted API Handling to a class which simplifies the task of understanding CorrectionProcessor, in doing so, we've established that InvocaApiCall should have public methods (`success?` and `unauthorized?`) that CorrectionProcessor can rely on. Future API updates can be accomplished so long as these methods still exist.
+  ```ruby
+  # Good
+    private
+
+    def invoca_api_call
+      @invoca_api_call ||= InvocaApiCall.new
+    end
+  end
+
+  class InvocaApiCall
+    def initialize
+      @response = JSON.parse(perform_api_call)
+    end
+
+    def perform_api_call
+      Net::Http.get('www.invoca.net')
+    end
+
+    def success?
+      @response["code"] == "200"
+    end
+
+    def unauthorized?
+      @response["code"] == "403"
+    end
+  end
+  ```
+
+</details>
+
+### Composition and Inheritance
+Generally, prefer Composing objects of behavior rather than relying on Inheriting behavior. Inheritance, especially after 2nd and 3rd levels of inheritance, becomes vastly harder to understand as it is extended.
+
+#### Composition Over Inheritance
+When faced with at least two (though some would argue three) examples of code reuse, extract that code reuse into a well-named module or object which represents that behavior.
```

Comment:
🔧 I'd also suggest adding a line that reminds the reader about `Single Responsibility Principle` in some way.

E.g.
```
A reminder that if a class is made up of a mix of many different abstractions (such as including many modules), this can also be a reminder to refactor so that the class's responsibilities do not continue to expand and grow too large.
```